netket/optimizer/sr/sr_onthefly_logic.py

```diff
@@ -99,78 +102,69 @@ def O_vjp(x, params, v, forward_fn, return_vjp_fun=False, vjp_fun=None, allreduc
     r"""
     compute  O^H w
     (where ^H is the hermitian transpose)
     """
+
+    if cast:
+        # determine the output type of the forward pass, because
+        # we need to cast the vectors in vjp to that type
+        dtype = jax.eval_shape(forward_fn, params, samples).dtype
+        w = w.astype(dtype)
+
     # O^H w = (w^H O)^H
     # The transposition of the 1D arrays is omitted in the implementation:
     # (w^H O)^H -> (w* O)*
 
     # TODO The allreduce in O_vjp could be deferred until after the tree_cast
     # where the amount of data to be transferred would potentially be smaller
-    w̄ = jnp.asarray(w.conjugate(), dtype=dtype)
+    res = tree_conj(O_vjp(samples, params, w.conjugate(), forward_fn, **kwargs))
 
-    res = tree_conj(O_vjp(samples, params, w̄, forward_fn, **kwargs))
-    #
     return tree_cast(res, params)
 
 
-def Odagger_DeltaO_v(samples, params, v, forward_fn, vjp_fun=None):
+def Odagger_O_v(samples, params, v, forward_fn, *, vjp_fun=None, center=False):
     r"""
-    compute \langle O^\dagger \Delta O \rangle v
-    where \Delta O = O - \langle O \rangle
+    if center=False (default):
+        compute \langle O^\dagger O \rangle v
 
-    optional: pass jvp_fun to be reused
-    """
+    else (center=True):
+        compute \langle O^\dagger \Delta O \rangle v
+        where \Delta O = O - \langle O \rangle
 
-    # determine the output type of the forward pass, because
-    # we need to cast the vectors in vjp to that type
-    dtype = jax.eval_shape(forward_fn, params, samples).dtype
+    optional: pass vjp_fun to be reused
+    """
 
-    # here the allreduce is deferred until after the dot product,
-    # where only scalars instead of vectors have to be summed
-    # the vjp_fun is returned so that it can be reused in OH_w below
-    omean, vjp_fun = O_mean(
-        samples,
-        params,
-        forward_fn,
-        dtype=dtype,
-        return_vjp_fun=True,
-        vjp_fun=vjp_fun,
-        allreduce=False,
-    )
-    omeanv = tree_dot(omean, v)  # omeanv = omean.dot(v); is a scalar
-    omeanv = sum_inplace(omeanv)  # MPI Allreduce w/ MPI_SUM
+    # w is an array of size n_samples; each MPI rank has its own slice
+    w = O_jvp(samples, params, v, forward_fn)
+    # w /= n_samples (elementwise):
+    w = w * (1.0 / (samples.shape[0] * n_nodes))
 
-    # v_tilde is an array of size n_samples; each MPI rank has its own slice
-    v_tilde = O_jvp(samples, params, v, forward_fn)
-    # v_tilde -= omeanv (elementwise):
-    v_tilde = v_tilde - omeanv
-    # v_tilde /= n_samples (elementwise):
-    v_tilde = v_tilde * (1.0 / (samples.shape[0] * n_nodes))
+    if center:
+        w = subtract_mean(w)  # w/ MPI
 
-    return OH_w(samples, params, v_tilde, forward_fn, dtype=dtype, vjp_fun=vjp_fun)
+    # TODO cast=True might be needed when the input and output are of different precision
+    return OH_w(samples, params, w, forward_fn, vjp_fun=vjp_fun, cast=False)
```

Comment:
@PhilipVinc Could you provide an example/explanation when the issue described in 4fcdd31091d352ced93231a597a8c0a86487a021 popped up?
I would like to add this to the test and I was assuming it's the case when the output of forward_fn is lower precision than the params, but I cannot seem to reproduce it.

Reply:
The problem arose because the first call to vjp was done in `O_mean`, where you generate the right sensitivity type because you use `abstraact_eval` to get the output dtype, but then `v_tilde` would become a different type so a second call to vjp, in `OH_w(samples, params, v_tilde, forward_fn, vjp_fun=vjp_fun)`, would fail.

I should have written the test case... but I had too much to do. I'm trying to find it now.
I think I found it while working on the tutorials..

Reply:
Found it 
It's this notebook: https://netket.github.io/netket/nk3/tutorials/j1j2.html (`tutorials/j1j2.ipynb` in the repo.) 

After reverting commit 4fcdd31091d352ced93231a597a8c0a86487a021 , and setting the default of center in sr_onthelfy to False to find it, I see the following error in cell 10:

```python
~/Documents/pythonenvs/netket_env/lib/python3.8/site-packages/jax/_src/scipy/sparse/linalg.py in cg(A, b, x0, tol, atol, maxiter, M)
    277   """
--> 278   return _isolve(_cg_solve,
    279                  A=A, b=b, x0=x0, tol=tol, atol=atol,

~/Documents/pythonenvs/netket_env/lib/python3.8/site-packages/jax/_src/scipy/sparse/linalg.py in _isolve(_isolve_solve, A, b, x0, tol, atol, maxiter, M, check_symmetric)
    217     if check_symmetric else False
--> 218   x = lax.custom_linear_solve(
    219       A, b, solve=isolve_solve, transpose_solve=isolve_solve,

~/Dropbox/Ricerca/Codes/Python/netket/netket/optimizer/sr/sr_onthefly_logic.py in mat_vec(v, forward_fn, params, samples, diag_shift, centered)
    234         f = Odagger_DeltaO_v
--> 235     res = f(samples, params, v, forward_fn)
    236     # add diagonal shift:

~/Dropbox/Ricerca/Codes/Python/netket/netket/optimizer/sr/sr_onthefly_logic.py in Odagger_DeltaO_v(samples, params, v, forward_fn, vjp_fun)
    165 
--> 166     return OH_w(samples, params, v_tilde, forward_fn, vjp_fun=vjp_fun)
    167 

~/Dropbox/Ricerca/Codes/Python/netket/netket/optimizer/sr/sr_onthefly_logic.py in OH_w(samples, params, w, forward_fn, **kwargs)
    122     # where the amount of data to be transferred would potentially be smaller
--> 123     res = tree_conj(O_vjp(samples, params, w.conjugate(), forward_fn, **kwargs))
    124     #

~/Dropbox/Ricerca/Codes/Python/netket/netket/optimizer/sr/sr_onthefly_logic.py in O_vjp(x, params, v, forward_fn, return_vjp_fun, vjp_fun, allreduce)
     90         _, vjp_fun = jax.vjp(forward_fn, params, x)
---> 91     res, _ = vjp_fun(v)
     92 

FilteredStackTrace: TypeError: Type of cotangent input to vjp pullback function (complex64) is not the expected tangent type (complex128) of corresponding primal output with dtype complex128.

The stack trace above excludes JAX-internal frames.
The following is the original exception that occurred, unmodified.
```

Reply:
I suspect this happens when model: `complex128 weights -> complex output`, sampler (therefore inputs) `float32`, hamiltonian dtype `float64`. 

because of float32 input, the output is complex64 and that tripped up the vjp

Reply:
let me know if that is clear enough

Reply:
I was able reproduce it using the the old code, thanks!

Running the new test on the old code also seems to detect it quite well, and also detects a few more test cases which didn't work even after https://github.com/netket/netket/commit/4fcdd31091d352ced93231a597a8c0a86487a021.

However the rewrite in this PR passes all of them. And casting is also is not necessary.
The notebook also runs fine with centered=False.